src/latedeliveryandpenalty-currency-conversion/model/clause.cto

```diff
@@ -0,0 +1,128 @@
+namespace org.accordproject.latedeliveryandpenalty
+
+import org.accordproject.cicero.contract.* from https://models.accordproject.org/cicero/contract.cto
+import org.accordproject.cicero.runtime.* from https://models.accordproject.org/cicero/runtime.cto
+import org.accordproject.time.* from https://models.accordproject.org/v2.0/time.cto
+import org.accordproject.money.CurrencyCode from https://models.accordproject.org/money@0.2.0.cto
+
+/**
+ * Defines the data model for the LateDeliveryAndPenalty template.
+ * This defines the structure of the abstract syntax tree that the parser for the template
+ * must generate from input source text.
+ */
+asset LateDeliveryAndPenaltyContract extends AccordContract {
+  /**
+   * The buyer
+   */
+  o AccordParty buyer
+
+  /**
+   * The seller
+   */
+  o AccordParty seller
+
+  /**
+   * Does the clause include a force majeure provision?
+   */
+  o Boolean forceMajeure
+
+  /**
+   * For every penaltyDuration that the goods are late
+   */
+  o Duration penaltyDuration
+
+  /**
+   * Seller pays the buyer penaltyPercentage % of the value of the goods
+   */
+  o Double penaltyPercentage
+
+  /**
+   * Up to capPercentage % of the value of the goods
+   */
+  o Double capPercentage
+
+  /**
+   * If the goods are >= termination late then the buyer may terminate the contract
+   */
+  o Duration termination
+
+  /**
+   * Fractional part of a ... is considered a whole ...
+   */
+  o TemporalUnit fractionalPart
+
+  /**
+   * Adding new currency related fields
+   */
+
+  /**
+   * Where are the currency versions coming from?
+   */
+  o String conversionSource
+
+  /**
+   * The currency that the goods are valued in
+   */
+  o CurrencyCode fromCurrency
+
+  /**
+   * The currency to be used for the penalty
+   */
+  o CurrencyCode toCurrency
+}
+
+/**
+ * Represents a currency conversion pair and exchange rate
+ */
+concept CurrencyConversion {
+   o CurrencyCode from
+   o CurrencyCode to
+   o Double rate
+}
+
+/**
+ * Defines the input data required by the template
+ */
+transaction LateDeliveryAndPenaltyRequest extends Request {
+
+  /**
+   * Are we in a force majeure situation? 
+   */
+  o Boolean forceMajeure
+
+  /**
+   * What was the agreed delivery date for the goods?
+   */
+  o DateTime agreedDelivery
+
+  /**
+   * If the goods have been delivered, when where they delivered?
+   */
+  o DateTime deliveredAt optional
+
+  /**
+   * What is the value of the goods?
+   */
+  o Double goodsValue
+
+  /**
+   * What is the exchange rate?
+   */
+  /*o Double exchangeValue*/
+  o CurrencyConversion currencyConversion
```

Comment:
If you make this a list then the "Oracle" can add as many currency conversions as it knows and it is up to the Ergo code to find the correct conversion to apply, based on the data in the contract.